Refactored online-tracker.component.ts to include a service (online-t…

jknowlesdev opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
online-tracker.component update


* **What is the current behavior?** (You can also link to an open issue here)
online-tracker.component.ts currently acts as an on/offline indicator as well as checking for internet connection.


* **What is the new behavior (if this is a feature change)?**
online-tracker.service.ts is created to act as a service for the online-tracker.component. The component now only acts as an on/offline indicator and relies on online-tracker.service to determine if there is connectivity.


* **Other information**:
This was done so we (Worcester team) can use the online-tracker-service for offline AMRS implementation.

fatmali commented:
Hi @jknowles-cs, Great job! I have some recommendations:

1. Fix the broken tests for the component
2. Write a unit test for the online tracker service.

Note: we always submit 1 commit per PR, so you can do a soft reset (**git reset HEAD~**), make the above changes, commit and force update. Before that, you can run tests using the command, **npm run test**, to make sure everything is okay.

enyachoke commented on src/app/online-tracker/online-tracker.service.spec.ts:
```diff
@@ -0,0 +1,78 @@
+
+  it('should return true when online', () => {
+    let service: OnlineTrackerService = TestBed.get(OnlineTrackerService);
+    this.service.isOnline = true;
```
@jknowles-cs  Why are calling this.service.isOnline? That should be service.isOnline.

enyachoke commented on src/app/online-tracker/online-tracker.component.html:
```diff
@@ -0,0 +1,6 @@
+<p *ngIf="this.onlineTrackerService.isOnline" class="text-bold"><i class="fa fa-circle text-success"></i>
+  <span></span>
+</p>
+<p *ngIf="!this.onlineTrackerService.isOnline" class="text-bold"><i class="fa fa-circle text-danger"></i>
```
You should move this to a helper method in the component which should then call the service

enyachoke commented on src/app/online-tracker/online-tracker.component.ts:
```diff
@@ -1,46 +1,14 @@
-    let timer = Observable.timer(1000, 30000);
-    timer.subscribe((t) => this.updateOnlineStatus());
+  constructor(private onlineTrackerService: OnlineTrackerService) {
+    this.onlineTrackerService.updateOnlineStatus();
```
You want to put this in the onInit function and reuse the helper  service I suggested above.

enyachoke commented on src/app/online-tracker/online-tracker.service.ts:
```diff
@@ -0,0 +1,36 @@
+
+  }
+
+  public ngOnInit() {
```
Rename this to avoid confusion.

jknowlesdev commented:
Hi everyone! Thank you all for your help! I have updated the code based on the suggestions provided by Felix (fmaiko) on Slack. Everything seems to be working!

I am still having trouble with the tests unfortunately. I would like to be able to test the getOnlineStatus() functions directly (in both the component and service) but I am unsure how to go about doing this.

Also, I have commented out most of the tests I am trying to perform. They are all failing besides the ones testing for creating an instance of the service and component. I am unsure of what's causing these tests to fail.

I appreciate the help and suggestions from everyone! I have learned a lot during this whole process.

Commit:
Refactored online-tracker.component to include a service (online-tracker-service)

This was done so we can use the online-tracker-service for offline AMRS implementation.

jknowlesdev commented:
I had to close this and reopen it here: [https://github.com/AMPATH/ng2-amrs/pull/656](https://github.com/AMPATH/ng2-amrs/pull/656)

The new PR is in its own branch instead of master.
